static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,382 @@
+		window?: WindowProxy;
+	},
+) => void;
+
+const LISTENERS: Listeners = {};
+let REGISTERED_LISTENERS = 0;
+
+const error405 = {
+	code: 405,
+	message: 'Service %% not implemented',
+};
+const error500 = {
+	code: 500,
+	message: 'Internal server error\n\n%%',
+};
+
+/**
+ * Determine if an unknown payload has the shape of a programmatic message
+ *
+ * @param payload The unknown message payload
+ */
+const isProgrammaticMessage = (
+	payload: unknown,
+): payload is ProgrammaticMessage => {
+	const payloadToCheck = payload as ProgrammaticMessage;
+	return (
+		payloadToCheck.type === 'set-ad-height' &&
+		('id' in payloadToCheck.value || 'slotId' in payloadToCheck.value) &&
+		'height' in payloadToCheck.value
+	);
+};
+
+/**
+ * Convert a legacy programmatic message to a standard message
+ *
+ * Note that this only applies to specific resize programmatic messages
+ * (these include specific width and height values)
+ */
+const toStandardMessage = (
+	payload: ProgrammaticMessage,
+): StandardMessage<{ width: number; height: number }> => ({
+	id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
+	type: 'resize',
+	iframeId: payload.value.id,
+	slotId: payload.value.slotId,
+	value: {
+		height: +payload.value.height,
+		width: +payload.value.width,
+	},
+});
+
+/**
+ * Retrieve a reference to the calling iframe via it's ID.
+ *
+ * Incoming messages contain the ID of the iframe into which the source window is embedded.
+ */
+const getIframe = (data: StandardMessage) => {
+	if (data.slotId) {
+		const container = document.getElementById(`dfp-ad--${data.slotId}`);
+		const iframes = container
+			? container.getElementsByTagName('iframe')
+			: null;
+		return iframes?.length ? iframes[0] : null;
+	} else if (data.iframeId) {
+		const el = document.getElementById(data.iframeId);
+		if (el instanceof HTMLIFrameElement) {
+			return el;
+		}
+		return null;
+	}
+};
+
+/**
+ * Convert an `unknown` payload to the standard message format
+ *
+ * Until DFP provides a way for us to identify with 100% certainty our
+ * in-house creatives, we are left with doing some basic tests
+ * such as validating the anatomy of the payload and whitelisting
+ * event type
+ */
+const isValidPayload = (payload: unknown): payload is StandardMessage => {
+	const payloadToCheck = payload as StandardMessage;
+	return (
+		'type' in payloadToCheck &&
+		'value' in payloadToCheck &&
+		'id' in payloadToCheck &&
+		payloadToCheck.type in LISTENERS &&
+		/^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/.test(payloadToCheck.id)
+	);
+};
+
+/**
+ * Cheap string formatting function
+ *
+ * @param error An object `{ code, message }`. `message` is a string where successive
+ * occurrences of %% will be replaced by the following arguments
+ * @param args Arguments that will replace %%
+ *
+ * @example
+ * formatError({ message: "%%, you are so %%" }, "Regis", "lovely")
+ * => { message: "Regis, you are so lovely" }
+ */
+const formatError = (error: { message: string }, ...args: string[]) =>
```

Comment:
This takes a `code`!

```suggestion
const formatError = (error: { code: number, message: string }, ...args: string[]) =>
```